File: impls/java-truffle/src/main/java/truffle/mal/step1_read_print.java
package truffle.mal;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class step1_read_print {

    public static void main(String[] args) throws IOException {
        boolean done = false;
        var reader = new BufferedReader(new InputStreamReader(System.in));
        while (!done) {
            System.out.print("user> ");
            String s = reader.readLine();
            if (s == null) {
                done = true;
            } else {
                try {
                    System.out.println(Printer.prStr(Reader.readStr(s), true));
                } catch (MalException ex) {
                    System.out.println(ex.getMessage());
                }
            }
        }
    }
}

File: impls/ruby.2/step1_read_print.rb
require "readline"

require_relative "errors"
require_relative "printer"
require_relative "reader"

module Mal
  extend self

  def READ(input)
    read_str(input)
  end

  def EVAL(input)
    input
  end

  def PRINT(input)
    pr_str(input, true)
  end

  def rep(input)
    PRINT(EVAL(READ(input)))
  rescue InvalidHashmapKeyError => e
    "Error! Hashmap keys can only be strings or keywords."
  rescue UnbalancedEscapingError => e
    "Error! Detected unbalanced escaping. Check for matching '\\'."
  rescue UnbalancedHashmapError => e
    "Error! Detected unbalanced list. Check for matching '}'."
  rescue UnbalancedListError => e
    "Error! Detected unbalanced list. Check for matching ')'."
  rescue UnbalancedStringError => e
    "Error! Detected unbalanced string. Check for matching '\"'."
  rescue UnbalancedVectorError => e
    "Error! Detected unbalanced list. Check for matching ']'."
  rescue SkipCommentError
    nil
  end
end

while input = Readline.readline("user> ")
  puts Mal.rep(input)
end

puts
